File: server/vehicle/commands.lua
local function deleteCurrentVehicle(ped)
    local entity = GetVehiclePedIsIn(ped, false)

    if entity and entity > 0 then
        local vehicle = Ox.GetVehicle(entity)

        if vehicle then
            vehicle:despawn()
        else
            DeleteEntity(entity)
        end
    end
end

lib.addCommand('car', {
    help = 'Spawn a vehicle using the given model',
    params = {
        { name = 'model', help = 'The model hash or archetype' },
        { name = 'owner', type = 'playerId', help = 'Add the vehicle to the database and set its owner as the given player', optional = true }
    },
    restricted = 'group.admin'
}, function(source, args)
    if args.owner and args.owner > 0 then
        args.owner = Ox.GetPlayer(args.owner)?.charId
    end

    local ped = GetPlayerPed(source)
    local vehicle = Ox.CreateVehicle({
        owner = args.owner,
        model = args.model,
    }, GetEntityCoords(ped), GetEntityHeading(ped))

    if vehicle then
        deleteCurrentVehicle(ped)

        for i = 1, 50 do
            Wait(0)
            SetPedIntoVehicle(ped, vehicle.entity, -1)

            if GetVehiclePedIsIn(ped, false) == vehicle.entity then
                break
            end
        end
    end
end)

lib.addCommand('dv', {
    help = 'Despawn your current vehicle, or vehicles within range',
    params = {
        { name = 'radius', type = 'number', help = 'Range to despawn vehicles, defaulting to 2', optional = true },
        { name = 'owned', type = 'string', help = 'Include player-owned vehicles', optional = true }
    },
    restricted = 'group.admin'
}, function(source, args)
    local ped = GetPlayerPed(source)

    if not args.radius then
        return deleteCurrentVehicle(ped)
    end

    local vehicles = lib.callback.await('ox:getNearbyVehicles', source, args.radius)

    for i = 1, #vehicles do
        local entity = NetworkGetEntityFromNetworkId(vehicles[i])
        local vehicle = Ox.GetVehicle(entity)

        if not vehicle then
            DeleteEntity(entity)
        elseif args.owned == 'true' then
            vehicle:despawn()
        end
    end
end)

File: server/vehicle/commands.ts
import { addCommand, triggerClientCallback } from '@overextended/ox_lib/server';
import { OxVehicle } from './class';
import { sleep } from '@overextended/ox_lib';
import { CreateVehicle } from 'vehicle';
import { OxPlayer } from 'player/class';

export function DeleteCurrentVehicle(ped: number) {
  const entity = GetVehiclePedIsIn(ped, false);

  if (!entity) return;

  const vehicle = OxVehicle.get(entity);

  vehicle ? vehicle.setStored('impound', true) : DeleteEntity(entity);
}

addCommand<{ model: string; owner?: number }>(
  'car',
  async (playerId, args, raw) => {
    const ped = playerId && GetPlayerPed(playerId as any);

    if (!ped) return;

    const player = args.owner ? OxPlayer.get(args.owner) : null;
    const data = {
      model: args.model,
      owner: player?.charId || undefined,
    };

    const vehicle = await CreateVehicle(data, GetEntityCoords(ped), GetEntityHeading(ped));

    if (!vehicle) return;

    DeleteCurrentVehicle(ped);
    await sleep(200);
    SetPedIntoVehicle(ped, vehicle.entity, -1);
  },
  {
    help: `Spawn a vehicle with the given model.`,
    params: [
      { name: 'model', paramType: 'string', help: 'The vehicle archetype.' },
      {
        name: 'owner',
        paramType: 'playerId',
        help: "Create a persistent vehicle owned by the target's active character.",
        optional: true,
      },
    ],
    restricted: 'group.admin',
  }
);

addCommand<{ radius?: number; owned?: string }>(
  'dv',
  async (playerId, args, raw) => {
    const ped = GetPlayerPed(playerId as any);

    if (!args.radius) return DeleteCurrentVehicle(ped);

    const vehicles = await triggerClientCallback<number[]>('ox:getNearbyVehicles', playerId, args.radius);

    if (!vehicles) return;

    vehicles.forEach((netId) => {
      const vehicle = OxVehicle.getFromNetId(netId);

      if (!vehicle) DeleteEntity(NetworkGetEntityFromNetworkId(netId));
      else if (args.owned) vehicle.setStored('impound', true);
    });
  },
  {
    help: `Deletes your current vehicle, or any vehicles within range.`,
    params: [
      { name: 'radius', paramType: 'number', help: 'The radius to despawn vehicles (defaults to 2).', optional: true },
      { name: 'owned', help: 'Include player-owned vehicles.', optional: true },
    ],
    restricted: 'group.admin',
  }
);
